pkg/wireguard/agent/agent.go

```diff
@@ -282,17 +284,26 @@ func (a *Agent) Init(ipcache *ipcache.IPCache, mtuConfig mtu.MTU) error {
 		return fmt.Errorf("failed to set link up: %w", err)
 	}
 
-	dev, err := a.wgClient.Device(types.IfaceName)
+	if err := a.restoreFromDevice(types.IfaceName); err != nil {
+		return err
+	}
+
+	// this is read by the defer statement above
+	addIPCacheListener = true
+
+	return nil
+}
+
+func (a *Agent) restoreFromDevice(name string) error {
+	dev, err := a.wgClient.Device(name)
 	if err != nil {
 		return fmt.Errorf("failed to obtain WireGuard device: %w", err)
 	}
+
 	for _, peer := range dev.Peers {
-		a.restoredPubKeys[peer.PublicKey] = struct{}{}
+		a.restoredPeers[peer.PublicKey] = peer.AllowedIPs
```

Comment:
One caveat here is that we may encounter issues in case one of the IPs previously allowed for one public key gets reused by another node (hence public key) before completing the ipcache synchronization. AFAIU, WireGuard enforces that any allowed IP is present only in one peer (as internally used for routing purposes), which means that the IP may incorrectly flip among two different nodes, causing connectivity problems. It is not totally clear to me whether this theoretical issue may also occur in practice, but I guess it is a possibility in case of ENI-like IPAMs (while with Kubernetes/cluster-pool we are guaranteed that a given pool is used by a single node). If this is confirmed, we may need a slightly more advanced logic to also handle that case, and remove the stale IP from the list in that case as well.

Reply:
Addressed in the latest version of the patch.